File: Remote/Source/Remote Elements/Views/ConnectionStatusButtonView.m
//
// ConnectionStatusButtonView.m
// Remote
//
// Created by Jason Cardwell on 5/24/11.
// Copyright 2011 Moondeer Studios. All rights reserved.
//
#import "RemoteElementView_Private.h"
#import "Button.h"
#import "ConnectionManager.h"

@interface ConnectionStatusButtonView ()
@property (nonatomic, strong) MSNotificationReceptionist * receptionist;
@end

@implementation ConnectionStatusButtonView

- (void)initializeIVARs {
  [super initializeIVARs];

  self.model.selected = [ConnectionManager isWifiAvailable];

  __weak ConnectionStatusButtonView * weakself = self;
  self.receptionist = [MSNotificationReceptionist
                       receptionistWithObserver:self
                                      forObject:[ConnectionManager class]
                               notificationName:CMConnectionStatusNotification
                                          queue:MainQueue
                                        handler:^(MSNotificationReceptionist * receptionist) {
                                          
                                          ConnectionStatusButtonView * view =
                                            (ConnectionStatusButtonView *)receptionist.observer;
                                          BOOL selected = view.model.selected;
                                          NSDictionary * userInfo = receptionist.notification.userInfo;
                                          NSNumber * value = userInfo[CMConnectionStatusWifiAvailableKey];
                                          BOOL wifiAvailable = BOOLValue(value);

                                          if (selected != wifiAvailable) view.model.selected = !selected;
                                        }];
}

@end

File: Remote/Source/Remote Elements/Views/ConnectionStatusButtonView.swift
//
//  ConnectionStatusButtonView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit

class ConnectionStatusButtonView: ButtonView {

  private var receptionist: MSNotificationReceptionist!

  /** initializeIVARs */
  override func initializeIVARs() {
    super.initializeIVARs()
    button.selected = ConnectionManager.isWifiAvailable()
    receptionist = MSNotificationReceptionist(
      observer: self,
      forObject: nil,
      notificationName: CMConnectionStatusNotification,
      queue: NSOperationQueue.mainQueue(),
      handler: {
        (receptionist: MSNotificationReceptionist!) -> Void in
          if let v = receptionist.observer as? ConnectionStatusButtonView {
            let selected = v.button.selected
            if let wifiAvailable = receptionist.notification.userInfo?[CMConnectionStatusWifiAvailableKey] as? Bool {
              if selected != wifiAvailable { v.button.selected = !selected }
            }
          }
      })
  }

}
